src/client/Microsoft.Identity.Client/OAuth2/MsalTokenResponse.cs

```diff
@@ -244,15 +244,28 @@ internal static MsalTokenResponse CreateFromManagedIdentityResponse(ManagedIdent
         {
             ValidateManagedIdentityResult(managedIdentityResponse);
 
+            long expiresIn = DateTimeHelpers.GetDurationFromNowInSeconds(managedIdentityResponse.ExpiresOn);

```

Comment:
Managed identity response provides only expires_**on**, not expires_**in**, right? The code sample in the issue #4062 is incorrect then?

Reply:
We map the managed identity response `ExpiresOn` to MSALTokenResponse `Expires_IN`. But some RPs do provide the actual Expires_In value, like in this case IMDS. You can run this [helper service query](https://service.msidlab.com/MSIToken?uri=http%3A%2F%2F169.254.169.254%2Fmetadata%2Fidentity%2Foauth2%2Ftoken%3Fapi-version%3D2018-02-01%26resource%3Dhttps%3A%2F%2Fmanagement.azure.com&azureResource=VM
), to see IMDS does return expires_in

But it leads to the discussion of when we should infer the refresh_on? In IMDS should we use what the RPs provide? @neha-bhargava @bgavrilMS

Reply:
Since the ExpiresOn value is consistent in the response, we can use the same logic. Aside I don't think there will be any difference between us calculating the expiresIn and the one returned from IMDS.

Reply:
Also I think for consistency it is better for us to infer